grype/presenter/json/presenter.go

```diff
@@ -31,11 +37,25 @@ func NewPresenter(pb models.PresenterConfig) *Presenter {
 		context:          pb.Context,
 		appConfig:        pb.AppConfig,
 		dbStatus:         pb.DBStatus,
+		outputFilePath:   outputFilePath,
+		fs:               fs,
 	}
 }
 
 // Present creates a JSON-based reporting
-func (pres *Presenter) Present(output io.Writer) error {
+func (pres *Presenter) Present(defaultOutput io.Writer) error {
```

Comment:
unfortunately this is breaking the presenter abstraction by adding filesystem concerns.

Today the presenter acts as a two-way facade... the caller provides doesn't need to know about any format concerns (is this json, table, xml, etc) and the presenter only needs to write bytes to the writer, but does not need to know about the concerns of that writer (is it to a file, the screen, streaming to a socket, etc).

Reply:
I think ultimately we should probably be getting rid of the presenter abstraction and inverting this entire flow... so instead of providing a writer to an object, we should ask the object to return bytes. If this is the way to go, we should try to do this incrementally. A good approach to this would be to take the GetPresenter() function and adapt that to return the new object and write a short generalized helper function that takes a presenter and returns bytes.   Initially we'd leave all of the presenters alone and fix everything above the GetPresenter() call (maybe call it GetWriter instead since it would no longer return a presenter). Then over time we can swap all of the presenters to encoders/writers.

I haven't thought this entirely though, but these are some initial thoughts.

Reply:
@olivierboudet I can try out the above suggestion on your branch, as it would ultimately enable the multi -o feature (which would be awesome). Would you mind if I pushed a lot of updates to your branch? 

(Alternatively I could try and get these changes in another PR then rebase this, which isn't a problem, but I feel the rebase would be nasty and the features on this branch nicely motivate the need for the change.)

Reply:
Thanks for the review ! Of course you can push on my branch ☺️